generate/templates/manual/repository/statistics.cc

```diff
@@ -0,0 +1,1912 @@
+
+    OdbObjectsData::iterTreeInfo itTreeInfo {};
+    if ((itTreeInfo = calculateTreeStatistics(commitOidTree)) == m_odbObjectsData.trees.info.end()) {
+      return false;
+    }
+
+    // update biggestCheckouts data
+    OdbObjectsData::TreeInfoAndStats &treeInfoAndStats = itTreeInfo->second;
+    m_statistics.biggestCheckouts.numDirectories = std::max<size_t>(
+      m_statistics.biggestCheckouts.numDirectories, treeInfoAndStats.stats.numDirectories);
+    m_statistics.biggestCheckouts.totalFileSize = std::max<size_t>(
+      m_statistics.biggestCheckouts.totalFileSize, treeInfoAndStats.stats.totalFileSize);
+    m_statistics.biggestCheckouts.maxPathDepth = std::max<size_t>(
+      m_statistics.biggestCheckouts.maxPathDepth, treeInfoAndStats.stats.maxPathDepth);
+    m_statistics.biggestCheckouts.numFiles = std::max<size_t>(
+      m_statistics.biggestCheckouts.numFiles, treeInfoAndStats.stats.numFiles);
+    m_statistics.biggestCheckouts.maxPathLength = std::max<size_t>(
+      m_statistics.biggestCheckouts.maxPathLength, treeInfoAndStats.stats.maxPathLength);
+    m_statistics.biggestCheckouts.numSymlinks = std::max<size_t>(
+      m_statistics.biggestCheckouts.numSymlinks, treeInfoAndStats.stats.numSymlinks);
+    m_statistics.biggestCheckouts.numSubmodules = std::max<size_t>(
+      m_statistics.biggestCheckouts.numSubmodules, treeInfoAndStats.stats.numSubmodules);
+  }
+
+  return true;
+}
+
+/**
+ * RepoAnalysis::calculateTreeStatistics
+ * 
+ * Calculates tree statistics recursively, considering individual tree's statistics
+ * have already been calculated.
+ * Returns an iterator to the tree info container, or to end if something went wrong.
+ */
+OdbObjectsData::iterTreeInfo RepoAnalysis::calculateTreeStatistics(const std::string &oidTree)
+{
+  OdbObjectsData::iterTreeInfo itTreeInfo = m_odbObjectsData.trees.info.find(oidTree);
+  if (itTreeInfo == m_odbObjectsData.trees.info.end()) {
+    return itTreeInfo;
+  }
+
+  OdbObjectsData::TreeInfoAndStats &treeInfoAndStats = itTreeInfo->second;
+
+  // prune recursivity
+  if (treeInfoAndStats.statsDone) {
+    return itTreeInfo;
+  }
+
+  ++treeInfoAndStats.stats.numDirectories;
+  ++treeInfoAndStats.stats.maxPathDepth;
+  // the following partial statistics have also been calculated in previous stage with threads:
+  // - treeInfoAndStats.stats.numFiles
+  // - treeInfoAndStats.stats.maxPathLength
+  // - treeInfoAndStats.stats.numSymLinks
+  // - treeInfoAndStats.stats.numSubmodules
+
+  // totalFileSize
+  OdbObjectsData::iterBlobInfo itBlobInfo {};
+  for (auto &oidBlob : treeInfoAndStats.entryBlobs)
+  {
+    if ((itBlobInfo = m_odbObjectsData.blobs.info.find(oidBlob)) == m_odbObjectsData.blobs.info.end()) {
+      return m_odbObjectsData.trees.info.end(); // to let the caller know that something went wrong
+    }
+
+    treeInfoAndStats.stats.totalFileSize += itBlobInfo->second.size;
+  }
+
+  // recursively into subtrees
```

Comment:
Is there any chance of stack overflow here?

Reply:
The maximum number of recursive calls depend directly on the maximum path depth of the repository. For instance, the linux repository have a maximum path depth of 13, so I'd say the chances of stack overflow are pretty low here, but it was very good to have it double checked 👍 Thanks!